server.cpp

```diff
@@ -40,38 +45,133 @@ size_t read_complete(const char* buff, const error_code& err, size_t bytes)
     return found ? 0 : 1;
 }
 
-void write_log(const std::string& outfile, const std::string& log_message)
+std::string make_log_line(const std::string& message)
+{
+    return make_daytime_string() + " " + message + "\n";
+}
+
+void write_log(const std::string& outfile, const std::string& message)
 {
     if (!outfile.empty())
+        std::ofstream(outfile, std::ios::app) << make_log_line(message);
+    else
+        std::cout << make_log_line(message);
+}
+
+void send_string(tcp::socket& socket, const std::string& str)
+{
+    boost::asio::write(socket, boost::asio::buffer(str));
+}
+
+void send_index(tcp::socket& socket, DIR *dir, const std::string& path)
+{
+    std::string lines;
+
+    for (struct dirent *entry = readdir(dir); entry != NULL; entry = readdir(dir))
     {
-        std::ofstream fout(outfile, std::ios::app);
-        fout << log_message << "\n";
+        if (strcmp(".", entry->d_name) && strcmp("..", entry->d_name))
+        {
+            const std::string file_name = entry->d_name;
+
+            struct stat st;
+            if (stat((path + "/" + file_name).c_str(), &st) < 0)
+            {
+                lines = lines + "<tr><td>" + file_name + "</td><td>?</td><td>?</td></tr>";
+            }
+            else
+            {
+                const std::string file_date = ctime(&st.st_ctime);
+                lines = lines + "<tr><td><p><a href=\"" + file_name + "\">" + file_name + "</a></p></td><td>" + std::to_string(st.st_size) + "</td><td>" + file_date + "</td></tr>";
+            }
+        }
+    }
+
+    const std::string table_html ="<!DOCTYPE html> \
+        <html> \
+        <body> \
+        <table border=\"1\" cellspacing=\"0\" cellpadding=\"5\"> \
+        <tr><td>File name</td><td>File size</td><td>Last modification date</td></tr>" + lines + " \
+        </table> \
+        </body> \
+        </html>";
+
+    const std::string index =  "HTTP/1.1 200 OK\r\nContent-Type: text/html; charset=utf-8\r\n\r\n" + table_html;
+
+    send_string(socket, index);
+}
+
+void write_file(tcp::socket& socket, const std::string& request_path_file, const std::string& path)
```

Comment:
Error processing in this function is kinda erratic - exceptions from the first half of the function are processed elsewhere, outside of the function, while exceptions from the second half are processed here, inside the function. You should be consistent in error processing.

Same problem with `write_response` below.

Reply:
Also, exception messages should go to the log.